support partition pruning for partitionable table source. requires ht…

godfreyhe opened the pull request.

godfreyhe committed:
support partition pruning for partitionable table source. requires https://issues.apache.org/jira/browse/FLINK-3849

godfreyhe committed:
fix compile error

godfreyhe commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/FilterableTableSource.scala:
```diff
@@ -0,0 +1,44 @@
+  * A [[TableSource]] extending this interface is able to filter the fields of the return table.
+  *
+  */
+trait FilterableTableSource {
```
目前仅依赖 FLINK-3849 中的 interface

beyond1920 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/functions/ScalarFunction.scala:
```diff
@@ -61,6 +61,12 @@ abstract class ScalarFunction extends UserDefinedFunction {
+    * @return true if a scalar function is guaranteed to always return
+    *         the same result given the same arguments; true is assumed by default
+    */
+  def isDeterministic: Boolean = true
```
引入这个的作用是干吗？能举个例子嘛？是说CurrentTime类似这种的函数就是false？

godfreyhe replied:
partition pruning是确定性的，对于非Deterministic的UDF，不能用于partition pruning的。CurrentTime是nonDeterministic

beyond1920 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/util/PartitionPredicateExtractor.scala:
```diff
@@ -0,0 +1,56 @@
+    */
+  def extractPartitionPredicate(
+    predicate: Array[Expression],
+    partitionFieldNames: Array[String]): (Array[Expression], Array[Expression]) = {
```
能写下可以滤掉哪些嘛？比如or, and各种复合的predict？

beyond1920 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/util/PartitionPredicateExtractor.scala:
```diff
@@ -0,0 +1,56 @@
+        postOrderVisit(r, partitionFieldNames) && partitionFieldNames.contains(name)
+      case s@ScalarFunctionCall(scalarFunction, _) =>
+        // skip non-deterministic scalar function
+        postOrderVisit(s, partitionFieldNames) && scalarFunction.isDeterministic
```
这个是指啥？非确定性的scalaFunction是指啥？

beyond1920 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/FilterableTableSource.scala:
```diff
@@ -0,0 +1,44 @@
+    * @param predicate a filter expression that will be applied to fields to return.
+    * @return an unsupported predicate expression.
+    */
+  def setPredicate(predicate: Array[Expression]): Array[Expression]
```
这个predict数组和Calc RelNode中的condition是什么关系？就是condition平铺开，转成多个Expression？？不太明白

godfreyhe replied:
一个RexNode按AND拆分成多个子RexNode，分别转化为Expression

beyond1920 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/util/PartitionPruner.scala:
```diff
@@ -0,0 +1,262 @@
+    val typeFactory = new FlinkTypeFactory(RelDataTypeSystem.DEFAULT)
+    val relDataType = typeFactory.buildRowDataType(partitionFieldNames, partitionFieldTypes)
+    val predicateRexNode = convertPredicateToRexNode(newPredicate, relBuilder, relDataType)
+
```
返回的predicateRexNode有可能是null，这里是不是要先做个判断

godfreyhe replied:
需要做一个判断更保险

beyond1920 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/util/PartitionPruner.scala:
```diff
@@ -0,0 +1,262 @@
+    predicate.map(_.postOrderTransform(rule))
+  }
+
+  private def conjunct(exps: Array[Expression]): Option[Expression] = {
```
这个的用处是？把多个Expression合成一个？为啥多个Expression之间合的时候用And？也有可能是Or呀，输入的Expression数组是根据什么分的？

godfreyhe replied:
这个 pr 依赖于 另外一个 pr（FilterTableSource， https://github.com/apache/flink/pull/3166/files ）。FilterableTableSource的接口定义就是把一个Filter条件（RexNode）先做了一次conjunction（去掉AND，转化为多个子RexNode）操作，然后分别转化为Expression，所以存在多个Expression。这部分逻辑在FilterTableSource的pr中。convertPredicateToRexNode，resolveFields，conjunct 这几个方法在另外一个 pr 中也有，我到时候需要等待他commit后，在抽取成共有的。 现在加这几个仅仅为了编译通过。该加点注释

beyond1920 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/util/PartitionPruner.scala:
```diff
@@ -0,0 +1,262 @@
+  * open and close methods in [[org.apache.flink.table.functions.UserDefinedFunction]]
+  * must be evaluated with [[org.apache.flink.api.common.functions.RuntimeContext]]
+  */
+class ScalarFunctionValidator extends RexVisitorImpl[RexNode](true) {
```
这个地方没看明白，为啥不让Function中覆盖open和close？

godfreyhe replied:
UDF是Job真正运行起来后调用的，这时可以拿到RuntimeContext，可以初始化open/close函数。Partition pruning 是运行在优化阶段。没有 RuntimeContext。如果用户的UDF中带有open/close，这里没法执行open/close

beyond1920 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/FilterableTableSource.scala:
```diff
@@ -0,0 +1,44 @@
+  def getPredicate: Array[Expression]
+
+  /**
+    * @param predicate a filter expression that will be applied to fields to return.
```
注释里面为啥是a filter expression，输入参数明明是个数组嘛

godfreyhe replied:
这个是FilterableTableSource的pr定义的interface，这里仅仅是加入进来编译可以通过

beyond1920 commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/FilterableTableSource.scala:
```diff
@@ -0,0 +1,44 @@
+
+  /**
+    * @param predicate a filter expression that will be applied to fields to return.
+    * @return an unsupported predicate expression.
```
return的主食里面为啥也是an expression, 明明返回数组嘛？

docete commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/util/PartitionPredicateExtractor.scala:
```diff
@@ -0,0 +1,56 @@
+
+  def postOrderVisit(e: Expression, partitionFieldNames: Array[String]): Boolean = {
+    e.children.forall {
+      case u@UnresolvedFieldReference(name) =>
```
这里的UnresolvedFieldReference和ResolvedFieldReference 有啥区别？

godfreyhe replied:
UnresolvedFieldReference 是未经 vlidate（类似SQL Validation） 的，ResolvedFieldReference 是 validate 后的。ExpressionParser.parseExpression 的结果是 UnresolvedFieldReference的

docete commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/plan/util/PartitionPruner.scala:
```diff
@@ -0,0 +1,262 @@
+  * open and close methods in [[org.apache.flink.table.functions.UserDefinedFunction]]
+  * must be evaluated with [[org.apache.flink.api.common.functions.RuntimeContext]]
+  */
+class ScalarFunctionValidator extends RexVisitorImpl[RexNode](true) {
```
这里是不允许分区字段上的UDF定义open/close方法吗？为什么？

godfreyhe replied:
是分区字段，原因同上

docete commented on flink-libraries/flink-table/src/main/scala/org/apache/flink/table/sources/PartitionableTableSource.scala:
```diff
@@ -0,0 +1,154 @@
+  }
+
+  /** return an predicate expression that was set. */
+  override def getPredicate: Array[Expression] = {
```
这里为啥默认返回Array.empty? 没看到例子里有重载和调用过

godfreyhe replied:
这个应该是返回 applyPredicate 后的结果，这个是 FilterableTableSource 里定义的， 我也没看到他代码哪里用到

godfreyhe committed:
add some comments
when partitionPredicate or partitionRexNode are empty, PartitionPruner return allPartitions

godfreyhe committed:
return empty as default when getPredicate in PartitionableTableSource
